### ISSUE TITLE
Assertion: stack underflow

### ISSUE THREAD

**Issue by pgavlin** [dotnet/llilc#636]
This pops up during a couple of the new IL tests.

`Assertion failed: (size() > 0) && "stack underflow", file ..\..\..\..\..\tools\llilc\lib\Reader\readerir.cpp, line 51`

**Comment by richardlford** [comment]
I'll look at these.

**Comment by richardlford** [comment]
The two tests failing with this failure are:

* \JIT\Regression\CLR-x86-EJIT\V1-M11-Beta1\b44018\b44018.cmd
* \JIT\Regression\v4\dev10_804810\dev10_804810.cmd

The b44018 test has IL like this:

```
		.method private hidebysig static int32 Main() il managed
		{
			.entrypoint
			.maxstack  30
			.locals (int32 V_0)
			ldc.i4.0
			stloc.0
			br begin
		begin:
			ldloc.0
			brtrue do2
			ldc.i4 100
			ret
		unreachable:
			ldc.i8 0
			br do1
		do1:
			pop
			ldloc.0
			brfalse do2
			br begin
		do2:
			ldc.i8 1
			br do1
		}
```

The block at do1 gets its stack state from the unreachable block. 
This originally led me to conclude that we needed to process
unreachable blocks in order to get the stack state right,
but ECMA I.1.8.1.1 states


    "The verification algorithm shall simulate all possible control flow paths
    through the code and ensure that a valid stack state exists for every 
    reachable CIL instruction."

So it seems that processing the unreachable code is overkill.
(continued in next comment block).

**Comment by richardlford** [comment]
The dev10_804810 test has IL like this:

```
    IL_0000:  br.s       IL_0029
    IL_0002:  brtrue.s   IL_0005
    IL_0004:  ret
    IL_0005:  br.s       IL_002c
    IL_0007:  brfalse.s  IL_0012
    IL_0009:  br.s       IL_002f
    IL_000b:  callvirt   instance void TestCase::Activate()
    IL_0010:  br.s       IL_0018
    IL_0012:  ldarg.1
    IL_0013:  callvirt   instance void TestCase::Deactivate()
    IL_0018:  ldarg.3
    IL_0019:  brfalse.s  IL_0022
    IL_001b:  ldarg.1
    IL_001c:  callvirt   instance void TestCase::LightUp()
    IL_0021:  ret
    IL_0022:  ldarg.1
    IL_0023:  callvirt   instance void TestCase::DimOut()
    IL_0028:  ret
    IL_0029:  ldarg.1
    IL_002a:  br.s       IL_0002
    IL_002c:  ldarg.2
    IL_002d:  br.s       IL_0007
    IL_002f:  ldarg.1
    IL_0030:  br.s       IL_000b
```

In this case the code at IL_0002 is only reachable via the branch at IL_002a
from which it gets a stack stage with one slot. This is an illegal program
according to the ECMA III.1.7.5 since the stack at IL_0002 should be empty.

**Comment by richardlford** [comment]
Issues #390 and #408 discuss the issue of the order in which basic blocks are
processed. The conclusion of that discussion was that we should process blocks in
MSIL order. However that does not deal with our current problem.

The essential requirement from those issues was that the type of 
phi instructions (at least their machine representation) must be determined
at the time that the block is turned into LLVM. Operands could be added
to the phi instructions later but they could not be allowed to change
the type of the code.

The ECMA spec talks about exploring all possible paths through the
method but seems to imply that this is being done in MSIL order
as when it talks about merging block states it talks about "existing"
state (i.e. state determined by earlier code) and "new state"
(i.e. state determined by later code). 

The kinds of merges that changes representation are:

* int32 and native int. 
* float32 and float64, but strictly speaking on the stack these are both
  represented as type F, which must actually be float64 in order to
  not lose information when a float64 is loaded onto the stack. But 
  currently LLILC handles these as if we could keep float32 and float64
  on the stack.

When merging stack state with these, we should keep the first type
encountered.

So it seems that the constraint on blocks with lexically preceding
predecessors is that they must follow their lexically-least predecessor.

I propose a new block order. The order will be partially constrained by a 
_dependency_ relation which is defined as follows:

* A block with predecessors that lexically (i.e. in MSIL order) precede it
  is dependent on those lexically-preceding blocks. I think this
  could be weakened to make it just depend on the lexically-least
  predecessor, but to minimize differences from our current
  code I would not do this.
* A block with no lexically-preceding blocks is dependent on 
  the lexically-least of its predecessor (which would be the one to
  determine the type of the phi, if any).

One could implement this by using a ready-list. Initially only the
start block is on the ready list. We then loop, taking blocks off
the ready list. When we finish processing a block we remove it
from the list of dependencies of its successors. Any successor
block which has no remaining dependencies is put on the
ready list.

**Comment by richardlford** [comment]
@erozenfeld, @pgavlin, @AndyAyersMS, @russellhadley, what do you think of this proposal?

**Comment by richardlford** [comment]
I just talked to @erozenfeld  and he pointed out a case where it would be 
impossible to satisfy this constraint. With this graph:

```
A: ...;goto C:
B: ...; goto C;
C: ... goto B;
```

B is only reachable by the backward branch from C, so we would want
it to be dependent on C, but C itself has B as a predecessor so
would depend on it. Thus we have cycles in the dependency relation
and there is no order to satisfy it.

I'm going to next study the RyuJit code to see how they handle this.

**Comment by JosephTremoulet** [comment]
You mention

> I think this could be weakened to make it just depend on the lexically-least predecessor

and I notice the cyclic example wouldn't have a cycle under that rule.  This suggests something more like a priority queue than a ready-list:

```
  Q.enqueue(Entry);
  while (!Q.empty) {
    B = Q.popLeast();
    process(B);
    for ( S : successors(B) ) {
      Q.enqueue(S); // weight is MSIL offset
    }
  }
```
(but modified to avoid double-processing)

This would
 - visit all reachable blocks
 - visit only reachable blocks
 - in graphs where each reachable block is the target of at least one lexically forward edge, visit all reachable blocks in MSIL order

**Comment by JosephTremoulet** [comment]
> in graphs where each reachable block is the target of at least one reachable lexically forward edge, visit all reachable blocks in MSIL order

probably good to double-check my math.  I think it goes like:

**claim:** given a program where each reachable block is the target of at least one reachable lexically forward edge,
   forall *n*, the block visited in step *n* is the lexically-first reachable block not visited in the prior steps
**base case:** step 0: entry is reachable, lexically first, and visited first
**inductive step:** if B is the lexically first reachable block not visited in steps 0..*n*-1, then
  1. all reachable blocks that lexically precede B have been visited 
  2. since B is reachable, by the property of the program we assumed it has some reachable predecessor P which lexically precedes it
  3. given 1&2, we know P has been visited, and therefore B has been queued
  4. since B has been queued and is the lexically first unvisited reachable block, it must have the min weight of all queued blocks (only reachable blocks get queued), so it will be visited next.

**Comment by JosephTremoulet** [comment]
> This suggests something more like a priority queue than a ready-list

Sorry, that didn't make sense, I'm not sure there's a meaningful difference between those two.  What I meant was simply that we could queue an item whenever any of its predecessors is visited, rather than requiring all lexically preceding predecessors to be visited or requiring the lexically-first predecessor to be visited, on the grounds that this agrees with lexical order for "normal" inputs.

**Comment by erozenfeld** [comment]
With this algorithm we will visit A, then C, then B in Rich's example. But B's stack may modify PHI types in C so we still have a problem.

**Comment by JosephTremoulet** [comment]
> But B's stack may modify PHI types in C so we still have a problem

I thought we wanted to prefer the "first" incoming type, which in this case would be A which is lexically first.  Has that changed?

**Comment by erozenfeld** [comment]
Yes, you are right. It works for that example.
What about this one:
```
A: ...;goto C:
B: ...;goto D:
C: ...;goto D:
D: ...;goto B:
```
The order will be A, C, D, B. But lexically first predecessor of D is B, not C.

**Comment by JosephTremoulet** [comment]
Right, the algorithm I outlined doesn't guarantee visiting lexically-first-reachable-predecessor first in a method with blocks that are only reachable by lexical back-edges (B in your example).  All it guarantees for those cases is that some reachable predecessor will have been visited first.  It doesn't sound to me like the other jits have been worrying over picking lexically-first-reachable-predecessor in those cases, so I was hoping we wouldn't need to either.
It does guarantee lexical order for methods with no blocks reachable only by lexical backedges, which I think is what we expect to see outside of hand-written IL tests.
If we think we need to guarantee lexically-first-reachable-predecessor for methods with blocks that are only reachable by lexical back-edges, perhaps we could try Rich's suggestion of "just depend on the lexically-least [reachable] predecessor", which presumably we could identify with a pre-pass that identifies which blocks are reachable, though we'd still have to figure out which block fails to have its lexically-first-reachable-predecessor processed first in cases with cyclic dependencies like
```
 A: ...; goto D;
 B: ...; br.cond E; // else fall-through to C
 C: ...; br.cond B; // else fall-through to D
 D: ...; goto B;
 E: etc
```
where B and C are each other's lexically-first-reachable-predecessors.

**Comment by erozenfeld** [comment]
>> If we think we need to guarantee lexically-first-reachable-predecessor for methods with blocks that are only reachable by lexical back-edges

Since all of these examples are illegal according to ECMA-335, we need to understand what RyuJIT does to figure out what rules we need to match.
